applications/TrilinosApplication/custom_strategies/convergencecriterias/trilinos_mixed_vector_scalar_criteria.h

```diff
@@ -0,0 +1,326 @@
+     */
+    TrilinosMixedVectorScalarCriteria(
+        TDataType VectorRatioTolerance,
+        TDataType VectorAbsTolerance,
+        TDataType ScalarRatioTolerance,
+        TDataType ScalarAbsTolerance,
+        const Variable<array_1d<double,3>>& rVectorVariable,
+        const Variable<double>& rScalarVariable)
+        : ConvergenceCriteria< TSparseSpace, TDenseSpace >(),
+        mVectorRatioTolerance(VectorRatioTolerance),
+        mVectorAbsTolerance(VectorAbsTolerance),
+        mScalarRatioTolerance(ScalarRatioTolerance),
+        mScalarAbsTolerance(ScalarAbsTolerance),
+        mrVectorVariable(rVectorVariable),
+        mrScalarVariable(rScalarVariable)
+    {}
+
+    /// Destructor.
+    ~TrilinosMixedVectorScalarCriteria() override {}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /// Compute relative and absoute error.
+    /**
+     * @param rModelPart Reference to the ModelPart containing the fluid problem.
+     * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
+     * @param A System matrix (unused)
+     * @param Dx Vector of results (variations on nodal variables)
+     * @param b RHS vector (residual)
+     * @return true if convergence is achieved, false otherwise
+     */
+    bool PostCriteria(
+        ModelPart& rModelPart,
+        DofsArrayType& rDofSet,
+        const TSystemMatrixType& A,
+        const TSystemVectorType& Dx,
+        const TSystemVectorType& b) override
+    {
+        // If required, initialize the Epetra vector import
+        if(!mEpetraImportIsInitialized) {
+            InitializeEpetraImport(rDofSet, Dx);
+        }
+
+        // Check if we are solving for something
+        if (TSparseSpace::Size(Dx) != 0) {
+            // Initialize values
+            TDataType vector_solution_norm = 0.0;
+            TDataType scalar_solution_norm = 0.0;
+            TDataType vector_correction_norm = 0.0;
+            TDataType scalar_correction_norm = 0.0;
+
+            int n_dofs = rDofSet.size();
+            unsigned int n_vector_dof = 0;
+            unsigned int n_scalar_dof = 0;
+            const auto& r_data_comm = rModelPart.GetCommunicator().GetDataCommunicator();
+            const int rank = r_data_comm.Rank();
+
+            // Do the local Dx vector import
+            Epetra_Vector local_dx( mpDofImport->TargetMap() );
+            int i_err = local_dx.Import(Dx, *mpDofImport, Insert);
+            KRATOS_ERROR_IF_NOT(i_err == 0) << "Local Dx import failed!" << std::endl;
+
+            // Loop over Dofs
+#pragma omp parallel reduction(+:vector_solution_norm,scalar_solution_norm,vector_correction_norm,scalar_correction_norm,n_vector_dof,n_scalar_dof)
+            {
+                int dof_id;
+                TDataType dof_value;
+                TDataType dof_dx;
+
+#pragma omp for
+                for (int i = 0; i < n_dofs; i++) {
```

Comment:
That should be 
```c++
for (int i = 0; i < n_dofs; i+=OpenMPUtils::GetNumThreads()) 
```

Should not change the results but it makes the loop not to scale with mpi+openmp
It was wong in the original mpi version. My bad

Reply:
I'd avoid the use of the `OpenMPUtils` (the intention is to remove it in the long term).

Reply:
@RiccardoRossi said that Epetra-vectors are (unless enabled in CMake) not OMP-compatible / threadsafe
There have been discussions abt this already in #5371, specifically https://github.com/KratosMultiphysics/Kratos/pull/5371#discussion_r328939651
=> the conclusion was to not do any omp on trilinos things

Reply:
But we aren't using it inside the omp context are we?. Anyway I have a new version in the oven :)

Reply:
I don't think the OpenMP is a problem here. It is only a problem in the FEM assembly